File: android/swapl/app/src/main/java/app/swapl/features/profile/IdentityVerificationCard.kt
package app.swapl.features.profile

import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.VerifiedUser
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.compose.LifecycleResumeEffect
import androidx.lifecycle.viewModelScope
import androidx.compose.ui.res.stringResource
import app.swapl.R
import app.swapl.core.repository.ReferralReward
import app.swapl.core.repository.VerificationRepository
import app.swapl.core.repository.VerificationStatus
import app.swapl.design.components.SurfaceCard
import app.swapl.designtokens.SwaplSpacing
import dagger.hilt.android.lifecycle.HiltViewModel
import io.ktor.client.plugins.ResponseException
import kotlinx.coroutines.launch
import javax.inject.Inject

// "Verify your identity" card for AccountScreen — Didit hosted verification.
//
// Hidden unless GET /api/verification/status says the feature is enabled AND
// the user isn't verified yet (env-gated, never broken). Tapping it mints a
// hosted Didit session and opens it in a Custom Tab; when the user comes back
// (ON_RESUME) we re-poll status so the card disappears once approved.

@HiltViewModel
class IdentityVerificationViewModel @Inject constructor(
    private val repo: VerificationRepository,
) : ViewModel() {
    var status by mutableStateOf<VerificationStatus?>(null)
        private set
    var busy by mutableStateOf(false)
        private set
    var error by mutableStateOf<String?>(null)
        private set

    // Post-verify referral reward to surface once via a snackbar. Cleared by
    // the UI after it's shown so it doesn't re-fire on recomposition.
    var pendingReward by mutableStateOf<ReferralReward?>(null)
        private set

    // Only re-poll on resume when we actually sent the user out to Didit —
    // otherwise every screen entry would fetch twice.
    private var awaitingReturn = false

    fun load() {
        viewModelScope.launch {
            runCatching {
                val fresh = repo.status()
                status = fresh
                val reward = fresh.referralReward
                if (reward != null && reward.keys > 0) pendingReward = reward
            }
        }
    }

    fun consumeReward() {
        pendingReward = null
    }

    fun refreshAfterReturn() {
        if (!awaitingReturn) return
        awaitingReturn = false
        load()
    }

    fun start(openUrl: (String) -> Unit) {
        if (busy) return
        busy = true
        error = null
        viewModelScope.launch {
            try {
                val started = repo.createSession()
                val url = started.url
                when {
                    started.status == "approved" -> load()
                    url != null -> {
                        awaitingReturn = true
                        openUrl(url)
                    }
                    else -> error = "Verification is unavailable right now. Try again later."
                }
            } catch (t: ResponseException) {
                when (t.response.status.value) {
                    // Feature switched off server-side since the status fetch — hide.
                    503 -> status = null
                    429 -> error = "Too many attempts — try again later."
                    else -> error = "Could not start verification. Try again."
                }
            } catch (_: Throwable) {
                error = "Could not start verification. Try again."
            } finally {
                busy = false
            }
        }
    }
}

@Composable
fun IdentityVerificationCard(
    vm: IdentityVerificationViewModel = hiltViewModel(),
    onReward: (String) -> Unit = {},
) {
    val context = LocalContext.current
    LaunchedEffect(Unit) { vm.load() }
    LifecycleResumeEffect(Unit) {
        vm.refreshAfterReturn()
        onPauseOrDispose { }
    }

    // Post-verify referral toast: fire once when the status carries a paid
    // reward, even though the card itself hides on `verified`.
    val reward = vm.pendingReward
    val rewardNamed = stringResource(R.string.verify_referral_toast_named)
    val rewardPlain = stringResource(R.string.verify_referral_toast)
    LaunchedEffect(reward) {
        val r = reward ?: return@LaunchedEffect
        val msg = r.referrerName?.let { name -> rewardNamed.format(name, r.keys) }
            ?: rewardPlain.format(r.keys)
        onReward(msg)
        vm.consumeReward()
    }

    val status = vm.status ?: return
    if (!status.enabled || status.verified) return

    SurfaceCard(
        modifier = Modifier.clickable(enabled = !vm.busy) {
            vm.start { url ->
                CustomTabsIntent.Builder().build().launchUrl(context, Uri.parse(url))
            }
        },
    ) {
        Column {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(SwaplSpacing.s3),
            ) {
                Icon(
                    Icons.Default.VerifiedUser,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.primary,
                )
                Column {
                    Text(
                        when (status.status) {
                            "pending" -> "Finish verifying your identity"
                            "declined" -> "Verification didn't go through"
                            else -> "Verify your identity"
                        },
                        style = MaterialTheme.typography.titleLarge,
                    )
                    Text(
                        when (status.status) {
                            "pending" -> "Pick up where you left off — it takes about two minutes."
                            "declined" -> "You can try again with a clearer photo of your ID."
                            else -> "Get the ID-verified badge hosts trust. Takes about two minutes."
                        },
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }
            vm.error?.let {
                Spacer(Modifier.height(SwaplSpacing.s2))
                Text(it, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.error)
            }
        }
    }
}

File: ios/Swapl/Features/Profile/IdentityVerificationCard.swift
import AuthenticationServices
import SwiftUI
import UIKit
import SwaplDesignTokens

// "Verify your identity" card for AccountView — Didit hosted verification.
//
// Hidden unless GET /api/verification/status says the feature is enabled AND
// the user isn't verified yet (env-gated, never broken). Tapping it mints a
// hosted Didit session and opens it in ASWebAuthenticationSession; when the
// sheet closes (callback or manual dismiss) we re-poll status and refresh
// /api/me so the existing "ID verified" badge updates.
struct IdentityVerificationCard: View {
    @Environment(AuthService.self) private var auth
    @State private var status: VerificationStatus?
    @State private var busy = false
    @State private var error: String?
    @State private var webSession = HostedVerificationSession()
    @State private var rewardToast: ReferralReward?

    var body: some View {
        Group {
            if let status, status.enabled, !status.verified {
                card(status: status)
            }
        }
        .task { await loadStatus() }
        .overlay(alignment: .bottom) {
            if let reward = rewardToast {
                Text(rewardMessage(reward))
                    .font(.swaplBody(SwaplDesignSystem.FontSize.bodySmall, weight: .semibold))
                    .foregroundStyle(SwaplSemanticLight.card)
                    .padding(.horizontal, 18)
                    .padding(.vertical, 12)
                    .background(AirbnbPalette.text, in: Capsule())
                    .padding(.bottom, 24)
                    .shadow(radius: 12, y: 4)
                    .transition(.move(edge: .bottom).combined(with: .opacity))
                    .onTapGesture { withAnimation { rewardToast = nil } }
            }
        }
    }

    private func rewardMessage(_ reward: ReferralReward) -> String {
        if let name = reward.referrerName {
            return "Verified! \(name)'s invite just earned you \(reward.keys) Keys. 🔑"
        }
        return "Verified! You earned \(reward.keys) Keys for joining via a referral. 🔑"
    }

    private func card(status: VerificationStatus) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            Button {
                startVerification()
            } label: {
                HStack(spacing: 18) {
                    ZStack {
                        RoundedRectangle(cornerRadius: SwaplDesignSystem.CornerRadius.medium, style: .continuous)
                            .fill(SwaplSemanticLight.accent)
                        if busy {
                            ProgressView().tint(SwaplSemanticLight.primary)
                        } else {
                            Image(systemName: "checkmark.shield")
                                .font(.system(size: 28, weight: .semibold))
                                .foregroundStyle(SwaplSemanticLight.primary)
                        }
                    }
                    .frame(width: 64, height: 64)

                    VStack(alignment: .leading, spacing: 5) {
                        Text(title(for: status.status))
                            .font(.swaplDisplay(20, weight: .semibold))
                            .foregroundStyle(AirbnbPalette.text)
                        Text(subtitle(for: status.status))
                            .font(.swaplBody(SwaplDesignSystem.FontSize.bodySmall))
                            .foregroundStyle(AirbnbPalette.secondaryText)
                            .lineLimit(2)
                            .multilineTextAlignment(.leading)
                    }
                    Spacer()
                    Image(systemName: "chevron.right")
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundStyle(AirbnbPalette.secondaryText)
                }
                .padding(18)
            }
            .buttonStyle(.plain)
            .disabled(busy)

            if let error {
                Text(error)
                    .font(.swaplBody(SwaplDesignSystem.FontSize.caption, weight: .semibold))
                    .foregroundStyle(SwaplSemanticLight.destructive)
                    .padding(.horizontal, 18)
                    .padding(.bottom, 14)
            }
        }
        .background(SwaplSemanticLight.card, in: RoundedRectangle(cornerRadius: SwaplDesignSystem.CornerRadius.large, style: .continuous))
        .overlay {
            RoundedRectangle(cornerRadius: SwaplDesignSystem.CornerRadius.large, style: .continuous)
                .stroke(AirbnbPalette.hairline)
        }
    }

    private func title(for state: String) -> String {
        switch state {
        case "pending": return "Finish verifying your identity"
        case "declined": return "Verification didn't go through"
        default: return "Verify your identity"
        }
    }

    private func subtitle(for state: String) -> String {
        switch state {
        case "pending": return "Pick up where you left off — it takes about two minutes."
        case "declined": return "You can try again with a clearer photo of your ID."
        default: return "Get the ID-verified badge hosts trust. Takes about two minutes."
        }
    }

    private func loadStatus() async {
        let fresh = try? await VerificationRepository.shared.status()
        status = fresh
        // Post-verify referral toast: show once when the status carries a paid
        // reward, then auto-dismiss. Best-effort; nil reward shows nothing.
        if let reward = fresh?.referralReward, reward.keys > 0, rewardToast == nil {
            withAnimation { rewardToast = reward }
            Task {
                try? await Task.sleep(for: .seconds(5))
                withAnimation { rewardToast = nil }
            }
        }
    }

    private func startVerification() {
        guard !busy else { return }
        busy = true
        error = nil
        Task {
            defer { busy = false }
            do {
                let start = try await VerificationRepository.shared.createSession()
                if start.status == "approved" {
                    await finish()
                    return
                }
                guard let raw = start.url, let url = URL(string: raw) else {
                    error = "Verification is unavailable right now. Try again later."
                    return
                }
                await webSession.present(url: url)
                await finish()
            } catch APIClient.APIError.status(503, _) {
                // Feature switched off server-side since the status fetch — hide.
                status = nil
            } catch {
                self.error = error.localizedDescription
            }
        }
    }

    // After the hosted sheet closes: converge our row (the status route also
    // polls Didit when webhooks aren't configured) and refresh /api/me so the
    // profile badge flips without an app restart.
    private func finish() async {
        await loadStatus()
        await auth.refreshSession()
    }
}

// Hosted-page presenter. The Didit callback returns to
// https://<app>/dashboard?verification=done; on iOS 17.4+ we catch it with an
// https callback and auto-dismiss, on earlier versions the user closes the
// sheet manually ("Cancel") — either way the continuation resumes and the
// caller re-polls status.
@MainActor
final class HostedVerificationSession: NSObject, ASWebAuthenticationPresentationContextProviding {
    private var activeSession: ASWebAuthenticationSession?

    func present(url: URL) async {
        await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
            let completion: ASWebAuthenticationSession.CompletionHandler = { _, _ in
                continuation.resume()
            }
            let session: ASWebAuthenticationSession
            if #available(iOS 17.4, *), let host = APIClient.shared.baseURL.host {
                session = ASWebAuthenticationSession(
                    url: url,
                    callback: .https(host: host, path: "/dashboard"),
                    completionHandler: completion
                )
            } else {
                // The custom scheme never fires for an https callback; the user
                // dismisses the sheet and we land in the same completion.
                session = ASWebAuthenticationSession(
                    url: url,
                    callbackURLScheme: "swapl",
                    completionHandler: completion
                )
            }
            session.presentationContextProvider = self
            session.prefersEphemeralWebBrowserSession = false
            activeSession = session
            session.start()
        }
        activeSession = nil
    }

    nonisolated func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        MainActor.assumeIsolated {
            UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .flatMap(\.windows)
                .first(where: \.isKeyWindow) ?? ASPresentationAnchor()
        }
    }
}
